lisa/tests/kernel/scheduler/load_tracking.py

```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
           ``error_margin_pct```) are allowed
         """
         return self._test_behaviour('load', error_margin_pct, allowed_error_pct)
+
+class CPUMigrationBase(LoadTrackingBase):
+    """
+    Base class for migration-related load tracking tests
+
+    The idea here is to run several rt-app tasks and to have them pinned to
+    a single CPU for a single phase. They can change CPUs in a new phase,
+    and we can then inspect the CPU utilization - it should match the
+    sum of the utilization of all the tasks running on it.
+
+    **Design notes:**
+
+    Since we sum up the utilization of each task, make sure not to overload the
+    CPU - IOW, there should always be some idle cycles.
+
+    The code assumes all tasks have the same number of phases, and that those
+    phases are all aligned.
+    """
+
+    PHASE_DURATION_S = 3 * UTIL_AVG_CONVERGENCE_TIME_S
+    """
+    The duration of a single phase
+    """
+
+    @classmethod
+    def _run_rtapp(cls, te, res_dir, profile):
+        # Just do some validation on the profile
+        for name, task in profile.items():
+            for phase in task.phases:
+                if len(phase.cpus) != 1 :
+                    raise RuntimeError("Each phase must be tied to a single CPU. "
+                                       "Task \"{}\" violates this".format(name))
+
+        super()._run_rtapp(te, res_dir, profile)
+
+    def __init__(self, res_dir, plat_info, rtapp_profile):
+        super().__init__(res_dir, plat_info, rtapp_profile)
+
+        self.cpus = set()
+
+        self.reference_task = list(self.rtapp_profile.values())[0]
+        self.nr_phases = len(self.reference_task.phases)
+
+        for task in self.rtapp_profile.values():
+            for phase in task.phases:
+                self.cpus.update(phase.cpus)
+
+        self.phases_durations = [phase.duration_s
+                                 for phase in self.reference_task.phases]
+
+    def get_expected_cpu_util(self):
+        """
+        Get the per-phase average CPU utilization expected from the rtapp profile
+
+        :returns: A dict of the shape {cpu : {phase_id : expected_util}}
+        """
+        cpu_util = {cpu : {phase_id : 0 for phase_id in range(self.nr_phases)}
+                    for cpu in self.cpus}
+
+        for task in self.rtapp_profile.values():
+            for phase_id, phase in enumerate(task.phases):
+                cpu_util[phase.cpus[0]][phase_id] += UTIL_SCALE * (phase.duty_cycle_pct / 100)
```

Comment:
Should not be more safe as:
```python
+= (UTIL_SCALE * phase.duty_cycle_pct) / 100
```

Reply:
Welcome to the wonderful world of Python3 where `20 / 100` == `0.2` (`20 // 100` == `0`)

Reply:
Yes, I guess it was P3 magic... however, we should maybe keep consistency with the Linux kernel world we come from, and the interpretation above is what we use in kernel space... just in case we want to be more friendly with some "low-level guys" looking at our tests ;)

Reply:
Yeah I feel you, I hated the guts out of that division change when I started doing Python3 stuff. Thing is, you can't hide it all the times with parentheses, so might as well embrace it.